Convert Vesync to 3.X version of library

cdnninja opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This converts to a re-written version the the pyvesync library.  It is a large change to both the integration and the code library under it.   Opening in draft to start for awareness.     Opening as well to allow me to see merge conflicts as I go and resolve.    Collaboration to get this ready is appreciated! 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [X] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #144807 fixes #147715 fixes #145123 
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

cdnninja committed:
First commit to support beta library

cdnninja committed:
Small tweaks

cdnninja committed:
More async and odds and ends

cdnninja committed:
fix deprecated manual_mode

cdnninja committed:
Remove speed range, point to library for data.

cdnninja committed:
Additional Library Alignment

cdnninja committed:
Merge branch 'dev' into 3.xvesync

cdnninja committed:
ruff / merge issue / temp requirements map

cdnninja commented:
To do list:

1.  Migration for entities, binary sensor for example have key changes. 
2. Diagnostics.py needs update to reflect new structure.
3. map data for sensor.py
4. Night light level number.py doesn't work yet
5. Tests
6. Air quality reports "1" vs "excellent"

cdnninja committed:
Merge branch 'home-assistant:dev' into 3.xvesync

cdnninja committed:
Fix key mapping and align with binary sensor approach

cdnninja committed:
bump version

cdnninja committed:
Tests and firmware

cdnninja committed:
Handle login failures with catching exception, fix number

cdnninja committed:
Remove SKU mapping table, fix diangostics

cdnninja committed:
remove dev type constant mapping for light

cdnninja committed:
Sensor mapping

cdnninja committed:
fix select logic

cdnninja committed:
Merge branch 'dev' into 3.xvesync

cdnninja committed:
Remove comment

cdnninja committed:
Little Bit Closer on init tests?

cdnninja committed:
More test fixes

cdnninja commented:
That authentication change hasn't been merged into the library. So you would also need to point to a copy of that pull request.  For those not impacted by that auth change yes ready to test.

SapuSeven reviewed (changes requested):
I tested your changes with a Vital 100S Air Purifier and it mostly worked.
The display control doesn't work because the naming in the library changed.

All other features (setting fan speed, preset modes, filter life etc) were working perfectly.

I added review comments on what I needed to change to get it to work.

cdnninja committed:
Update fan.py

cdnninja commented on homeassistant/components/vesync/fan.py:
```diff
@@ -148,56 +139,54 @@ def extra_state_attributes(self) -> dict[str, Any]:
 
         if hasattr(self.device, "mode"):
-            attr["mode"] = self.device.mode
+            attr["mode"] = self.device.state.mode
```
Updated to self.device.state.    I haven't tested that yet but committed in the meantime.

cdnninja committed:
toggle_display

cdnninja committed:
toggle_display

cdnninja commented on homeassistant/components/vesync/switch.py:
```diff
@@ -48,8 +48,8 @@ class VeSyncSwitchEntityDescription(SwitchEntityDescription):
+        exists_fn=lambda device: rgetattr(device, "state.display_status") is not None,
         translation_key="display",
         on_fn=lambda device: device.turn_on_display(),
         off_fn=lambda device: device.turn_off_display(),
```
Fixed.

cdnninja committed:
display_status

cdnninja commented on homeassistant/components/vesync/fan.py:
```diff
@@ -148,56 +139,54 @@ def extra_state_attributes(self) -> dict[str, Any]:
 
         if hasattr(self.device, "screen_status"):
-            attr["screen_status"] = self.device.screen_status
+            attr["screen_status"] = self.device.state.screen_status
```
Fixed.

cdnninja commented:
@SapuSeven  Thank you for the review! I have made the majority of the changes. Are you able to resolve the items if they look correct now?

joostlek commented on homeassistant/components/vesync/binary_sensor.py:
```diff
@@ -31,20 +31,24 @@ class VeSyncBinarySensorEntityDescription(BinarySensorEntityDescription):
-        is_on=lambda device: device.details["water_tank_lifted"],
+        is_on=lambda device: device.state.water_tank_lifted,
         device_class=BinarySensorDeviceClass.PROBLEM,
+        exists_fn=lambda device: rgetattr(device, "state.water_tank_lifted")
+        is not None,
```
It would maybe be nice to phase out the usage of has attr and get attr

joostlek replied:
But we can test that a device is of type humdifier? As in, I'd rather do an ifinstance check and `if dev.sate.water_tank_lifted is not None` than do getattr, as its not typesafe

cdnninja replied:
rgetattr I think is typesafe, getattr isn't though.      For this item I could switch it assuming all humidifiers have this sensor available even if None.    Not sure that will always be true for all sensor, or binary sensors in the future.

joostlek commented on homeassistant/components/vesync/number.py:
```diff
@@ -28,22 +28,24 @@
+    exists_fn: Callable[[VeSyncBaseDevice], bool] = lambda _: True
     value_fn: Callable[[VeSyncBaseDevice], float]
-    set_value_fn: Callable[[VeSyncBaseDevice, float], bool]
+    native_min_value_lambda: Callable[[VeSyncBaseDevice], float]
+    native_max_value_lambda: Callable[[VeSyncBaseDevice], float]
```
we use `_fn` as suffix for all functions, so let's stay consistent :)

cdnninja replied:
Corrected.

joostlek commented on homeassistant/components/vesync/sensor.py:
```diff
@@ -252,6 +211,9 @@ def native_value(self) -> StateType:
 
-    def update(self) -> None:
+    async def async_update(self) -> None:
         """Run the update function defined for the sensor."""
-        return self.entity_description.update_fn(self.device)
+        # I think this section can be removed since energy updates are now auto called?
+        # if self.entity_description.update_fn:
+        # return await self.entity_description.update_fn(self.device)
+        return
```
I am actually also not sure why energy had that separate function, I think this is something we have to find out tho

cdnninja committed:
Bump API version

cdnninja committed:
state fix.

cdnninja committed:
Typo

SapuSeven commented on homeassistant/components/vesync/fan.py:
```diff
@@ -101,26 +94,24 @@ class VeSyncFanHA(VeSyncBaseEntity, FanEntity):
         ):
-            return ranged_value_to_percentage(
-                SPEED_RANGE[SKU_TO_BASE_DEVICE[self.device.device_type]], current_level
+            return ordered_list_item_to_percentage(
+                self.device.fan_levels, current_level
             )
```
Not sure when exactly it happens.

One way to reliably reproduce it:
1. Delete all VeSync devices
2. Set the purifier to "Auto" mode
3. Add the device in HomeAssistant
4. Try to set the manual fan speed immediately after adding the device

In general, the controls from homeassistant feel broken when this happens - e.g. after clicking a manual setting, the Auto preset is still selected and it's impossible to switch to Auto directly (need to select another preset first then back).
After adding the additional check I suggested it works better.

There's still an issue with the controls, but I don't know if we can fix that: Everything that's changed is sent and applied to the device, but the controls in HA don't update - only after a minute when the new state is polled.
Is it possible to "optimistically" set the state in HA immediately upon change?

cdnninja replied:
This also speaks to the same item I believe. I see it on my 131s for display state: https://github.com/webdjoe/pyvesync/issues/328

cdnninja replied:
Mine doesn't do this - I do have an issue that I am missing a speed though.  I should have three speeds but I only have off and two others.   

0 comes from the API only when in a "mode" which the if statement above should catch showing it is not in manual mode.   When this occurs for your device what mode does the API show it is in? 

```
        device = manager.devices.air_purifiers[0]
        print(device.fan_levels)
        print(device.state.fan_level)
        print(device.modes)
        print(device.state.mode)
```

Shows me:

```
[1, 2, 3]
0
[<PurifierModes.SLEEP: 'sleep'>, <PurifierModes.MANUAL: 'manual'>, <PurifierModes.AUTO: 'auto'>]
sleep
```

In my case since mode is sleep the if statement catches it.  I tried same for auto and works as intended.   I do thing something is up with the count of speeds though causing missing slow speed. 

Are you missing a speed in the list?

SapuSeven replied:
I also have 4 speeds on the device, but HA only shows OFF/LOW/HIGH. I think this is just a UI issue though.

The debug logs, added just before the condition, show me the following:
```
[1, 2, 3, 4]
0
[<PurifierModes.SLEEP: 'sleep'>, <PurifierModes.MANUAL: 'manual'>, <PurifierModes.AUTO: 'auto'>, <PurifierModes.PET: 'pet'>]
manual
```
`state.fan_level` stays at `0` even if I select something different.
Then, when the update comes in, it shows the correct value:
```
[1, 2, 3, 4]
2
[<PurifierModes.SLEEP: 'sleep'>, <PurifierModes.MANUAL: 'manual'>, <PurifierModes.AUTO: 'auto'>, <PurifierModes.PET: 'pet'>]
**manual**
```

SapuSeven replied:
Also I noticed the following: Being on fan speed 2, selecting AUTO changes the mode immediately:
```
[1, 2, 3, 4]
2
[<PurifierModes.SLEEP: 'sleep'>, <PurifierModes.MANUAL: 'manual'>, <PurifierModes.AUTO: 'auto'>, <PurifierModes.PET: 'pet'>]
auto
```
Then, clicking a different manual speed setting changes the mode immediately:
[1, 2, 3, 4]
2
[<PurifierModes.SLEEP: 'sleep'>, <PurifierModes.MANUAL: 'manual'>, <PurifierModes.AUTO: 'auto'>, <PurifierModes.PET: 'pet'>]
manual
```
...but the speed doesn't update, only after the next poll

cdnninja replied:
So it looks like this is a library issue rather than integration.  0 is correctly not displayed as it means the device is in a mode other than manual.   The library becomes inconsistent though when commands are used, sometimes updating the value automatically, other times waiting for refresh.   This gets us in a spot where it is 0 and manual which isn't possible.  I will add this to the existing open issue on the API repo.  

As for the interface missing a mode I found that bug.  Pushing a fix shortly.

SapuSeven replied:
I figured it out.
The state has two properties: `fan_level` and `fan_set_level`. It becomes relevant when using auto/sleep modes:
The former indicates the actual fan speed, the latter remains constant when set to auto/sleep.
`fan_set_level` is the value we actually control by setting the fan speed, so it should be the one displayed in the HA control.
Fixed in a38dc2c4723fdf7c7f88321975e0fe1629f00e0b, I'll propose renaming the properties in pyvesync to make this clearer.

cdnninja commented on homeassistant/components/vesync/binary_sensor.py:
```diff
@@ -31,20 +31,24 @@ class VeSyncBinarySensorEntityDescription(BinarySensorEntityDescription):
     ),
     VeSyncBinarySensorEntityDescription(
-        key="details.water_tank_lifted",
+        key="state.water_tank_lifted",
```
Corrected.

cdnninja committed:
function name correction

SapuSeven commented:
@cdnninja all my review comments (except the one I just replied to) are resolved, but I don't seem to have permissions to mark them as resolved...

I was able to add and control my purifier, with two exceptions:
- _"No module named 'deprecated'"_ error on adding the device
- Air quality level is always `None`

Both of those are issues in `pyvesync` and will be fixed with https://github.com/webdjoe/pyvesync/pull/342/

cdnninja commented:
@SapuSeven  any chance CI tests are something you could take a stab at on this?  Beyond my python knowledge and have Faught them for a bit, I am not sure how to switch them over correctly for aiohttp.

cdnninja committed:
Clean up coordinator

cdnninja commented on homeassistant/components/vesync/coordinator.py:
```diff
@@ -38,12 +38,10 @@ def __init__(
-        self._manager.update_all_devices()
-        # Vesync updates energy on applicable devices every 6 hours
-        self._manager.update_energy()
+        await self._manager.update_all_devices()
```
Completed.

joostlek commented:
Ideally we patch on library level though, as in, we shouldn't use http mocks.

But the requests -> aiohttp mock change is probably the easiest for this PR

SapuSeven commented on homeassistant/components/vesync/__init__.py:
```diff
@@ -69,28 +61,19 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> b
     async def async_new_device_discovery(service: ServiceCall) -> None:
         """Discover if new devices should be added."""
+        # needs to be corrected.
         manager = hass.data[DOMAIN][VS_MANAGER]
         devices = hass.data[DOMAIN][VS_DEVICES]
 
-        new_devices = await async_generate_device_list(hass, manager)
+        new_devices = manager.devices
```
How exactly is this supposed to be corrected?
What does/should it do?

cdnninja committed:
Fix missing speeds in fan

SapuSeven committed:
Fix fan level not updating in HA when changed

cdnninja commented:
> > @spdustin could you retest with the latest?
> 
> Integration loaded and found my 100S (Levoit badged).
> 
> ### Issues
> 1. Entities get proper friendly names, but their entity IDs are (mostly) wonky.
> 2. Controlling an entity doesn't result in a logbook entry.
>    
>    * It _was_ adding logbook entries after my first attempt, but that's when I realized I had forgotten the translations file; once I removed the integration, added the translations file (copying `strings.json` to `translations/en.json`), rebooted, and activated the integration again, it stopped adding logbook entries.
> 
> ### Entities (`id` "Friendly Name")
> #### Controls
> * `fan.vital_100s` "Vital 100S"
>   
>   * The presets listed are `Pet`, `Sleep`, and `[%key:common::state::auto%]`. The latter immediately goes to 100%
> * `switch.vital_100s_none` "Display"
>   
>   * It works in that it updates the device when I switch it. However, the switch in HA immediately goes back to its previous value, waits a bit, and then updates to the "correct" value. If this relies on polling, perhaps an "optimistic" mode for this switch entity that waits a bit after it's been changed before polling for its current value?
> 
> #### Sensors
> * `sensor.vital_100s_pm2_5` "PM2.5"
>   
>   * Has a value of "1 µg/m³", though I suspect that isn't a "real" measurement, unless the PM sensor in these filters is legit. It feels more like an arbitrarily scaled value, as its "native" prop name is `air_quality_value`.
> * `sensor.vital_100s_none_2` "Air quality"
>   
>   * Has a value of "None", seems like the text value of the `air_quality_string` prop
> 
> #### Diagnostic
> * `sensor.vital_100s_none` "Filter lifetime"
>   
>   * Has a value of "99%", which matches the returned `filter_life` prop
> 
> Diagnostic Info

Thank you for testing and all the details! 

The screen display looks like this issue: https://github.com/webdjoe/pyvesync/issues/328

The text for air quality I opened this: https://github.com/webdjoe/pyvesync/issues/344

The preset list I suspect is something to do with the translation file.  I will poke around and try replicate that.   With the core install of the integration it seems to work okay with auto. 

Beyond that looks like working okay? I assume the sensor name of _none was from the translation files.

spdustin commented:
> Beyond that looks like working okay? I assume the sensor name of _none was from the translation files.

Beyond that, yeah, so far. 👍 I'd like to see entities for the other props surfaced in the component, but that's a v.Next thing, heh.

I'm currently going through the tests to see if I can help out there. It's going…slowly

cdnninja commented:
> > Beyond that looks like working okay? I assume the sensor name of _none was from the translation files.
> 
> Beyond that, yeah, so far. 👍 I'd like to see entities for the other props surfaced in the component, but that's a v.Next thing, heh.
> 
> I'm currently going through the tests to see if I can help out there. It's going…slowly

Yes this is a big PR as is, the home assistant team likes small ones.    Once this one is merged and the new API released adding sensors and switches should be pretty easy.   

Any help is appreciated!   @SapuSeven @spdustin  we have a small group chat on discord I can add you to if interested so we can coordinate efforts.

SapuSeven committed:
Update pyvesync version

SapuSeven committed:
Add @SapuSeven as a code owner for VeSync

cdnninja committed:
Adapt tests for pyvesync v3 (#639)

* Fix login response handling

* Start fixing aio-mocks for tests

* update 131 details

* Fix more http mocks for tests

* Update more tests

* Fix "select" tests and humidifier-300s fixture

* Fix "platform" tests and fix outlet sensor entities

* Cleanup air purifier fixture models

* Properly update energy history sensors

* Fix humidifier tests

* Fix device-detail response fixture

* Update device fixtures and mocks

* Fix diagnostics tests

* Fix switch tests

* Fix sensor tests

* Fix fan tests

* Fix config_flow tests

* Fix light tests

* Prepare for https://github.com/webdjoe/pyvesync/pull/348

* Fix init test

* Simplify login error handling for configured instance

* Ensure night_light entity is only created once

---------

Co-authored-by: SapuSeven <sapuseven@gmail.com>

SapuSeven committed:
Create separate entities for fan/outlet nightlight

cdnninja committed:
Fix login method

cdnninja committed:
Update for deprecated method

cdnninja committed:
Pass AIO session

cdnninja committed:
Bump version and adjust fan based on library change

SapuSeven committed:
Fix display state not updating immediately

SapuSeven committed:
Replace aioresponses with aioclient_mock

cdnninja committed:
Add session to config flow

cdnninja committed:
Fix fan not switching off mode

cdnninja committed:
Try fix diagnostics test

cdnninja committed:
Revert μg/m³

cdnninja committed:
Discovery Code Update

cdnninja committed:
Merge branch 'dev' into 3.xvesync

cdnninja committed:
Fix Tests due to encoding issue

cdnninja committed:
Bump API and add details to errors

cdnninja committed:
fix switch message

cdnninja commented:
Two outstanding items I am aware of - I am sure more bugs will be found and fixed beyond that.   However I think ready for review.  I would like this to be targeted for the 2025.10 release if we can.  

To do before merge:

1. Non RC library release. 
2. One bug fix for login pending merge in that library. 
3. Number entity has comments and needs some manual testing.

cdnninja committed:
Align set methods

SapuSeven committed:
Improve logging

- Log detailed login errors in config flow
- Remove redundant "An error occurred..." for most HomeAssistantErrors - they are already prefixed with "Error during service call to example/turn_off:"
- Add space for remaining exception messages for proper formatting after concatenation

cdnninja commented on homeassistant/components/vesync/fan.py:
```diff
@@ -191,7 +191,7 @@ async def async_set_percentage(self, percentage: int) -> None:
         ):
             raise HomeAssistantError(
-                "An error occurred while changing fan speed:"
+                "An error occurred while changing fan speed: "
```
I see in other items removing the context of the sentence and just putting the response message.   Is that intentional to have this one different?

SapuSeven replied:
Intentional, thought it would be helpful distinguish between the different actions within `async_set_percentage` (Turn on/off, set speed, manual mode)

cdnninja committed:
Remove comment align number

cdnninja committed:
Align response to smart tower fan actual

SapuSeven committed:
Update pyvesync version to latest beta (3.0.0b8)

joostlek committed:
Merge branch 'dev' into 3.xvesync

joostlek committed:
fix

cdnninja committed:
Fix light test

cdnninja committed:
Merge branch 'dev' into 3.xvesync
